src/snowflake/snowpark/_internal/ast/builder.py

```diff
@@ -0,0 +1,159 @@
+    and to generate AST requests.
+
+    The core statement types are:
+    - Bind: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    __generate_request_id = uuid.uuid4
+
+    # Class variable for generating IDs for bind statements.
+    __id_gen = itertools.count(start=1)
+
+    # NOTE: itertools.count and its __next__ method are only thread-safe in CPython with GIL.
+    # In other implementations (Jython, PyPy), this may not be thread-safe, and once the GIL is removed from CPython, it will not be thread-safe either.
+    # Likely, it would be best to make all of the methods in this class thread-safe with threading.Rlock
+    def __init__(
+        self,
+    ) -> None:
+        """
+        Initializes a new AST batch.
+        """
+        self._bind: proto.Bind = None
+        self.__eval: proto.Eval = None
+        self.__dependencies: List[AstBuilder] = []
+        self.__required_by: List[AstBuilder] = []
+
+    @classmethod
+    def Bind(cls, symbol: Optional[str] = None) -> "AstBuilder":
+        """
+        Factory method to create an AstBuilder instance with a new bind statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable in the bind statement.
+        Returns:
+            An instance of AstBuilder with a bind statement initialized.
+        """
+        b = cls()
+        b._bind = proto.Bind()
+        b._bind.uid = next(AstBuilder.__id_gen)
+        b._bind.symbol.value = symbol if isinstance(symbol, str) else ""
+        return b
+
+    def depends_on(self, dependency: "AstBuilder") -> None:
+        """
+        Adds a dependency to the current AstBuilder instance.
+        This allows for tracking dependencies between different AST statements.
+
+        Args:
+            dependency: An AstBuilder instance that is a dependency of the current instance.
+        """
+        if dependency not in self.__dependencies:
+            self.__dependencies.append(dependency)
+        if self not in dependency.__required_by:
+            dependency.__required_by.append(self)
+
+    @staticmethod
+    def _closure(root: "AstBuilder") -> List["AstBuilder"]:
+        """
+        Performs a post-order traversal of the AST to compute the transitive closure of dependencies.
+        This ensures that all dependencies are included in the final AST request.
+
+        Args:
+            root: The root AstBuilder instance from which to compute the closure.
+        """
+        q: deque[AstBuilder] = deque()
+        q.append(root)
+
+        visited = set()
+        closure = []
+
+        while q:
+            ast_builder = q.pop()
+            closure.append(ast_builder)
+            visited.add(ast_builder)
+            for dep in ast_builder.__dependencies:
+                if dep not in visited:
+                    q.append(dep)
+        closure.reverse()
```

Comment:
nit: this code is fine with the reverse, alternative is to prepend instead of append.